src/container/actions.ts

```diff
@@ -14,57 +14,84 @@ export const onDragNode: IOnDragNode = ({ config, event, data, id }) => (chart:
   const nodechart = chart.nodes[id]
 
   if (nodechart) {
-    chart.nodes[id] = {
+    const chartToUpdate = {
       ...nodechart,
-      position: config && config.snapToGrid ? { x: Math.round(data.x / 20) * 20, y: Math.round(data.y / 20) * 20 } : data,
-    }
+      position:
+        config && config.snapToGrid
+          ? { x: Math.round(data.x / 20) * 20, y: Math.round(data.y / 20) * 20 }
+          : data
+    };
+    return { ...chart, nodes: { ...chart.nodes, [id]: chartToUpdate } };
   }
 
   return chart
 }
 
 export const onDragCanvas: IOnDragCanvas = ({ config, event, data }) => (chart: IChart): IChart => {
-  chart.offset = config && config.snapToGrid ? { x: Math.round(data.x / 20) * 20, y: Math.round(data.y / 20) * 20 } : data
-  return chart
+  return {
+    ...chart,
+    offset:
+      config && config.snapToGrid
+        ? { x: Math.round(data.x / 20) * 20, y: Math.round(data.y / 20) * 20 }
+        : data
+  };
 }
 
 export const onLinkStart: IOnLinkStart = ({ linkId, fromNodeId, fromPortId }) => (chart: IChart): IChart => {
-  chart.links[linkId] = {
-    id: linkId,
-    from: {
-      nodeId: fromNodeId,
-      portId: fromPortId,
-    },
-    to: {},
-  }
-  return chart
+  return {
+    ...chart,
+    links: {
+      ...chart.links,
+      [linkId]: {
+        id: linkId,
+        from: {
+          nodeId: fromNodeId,
+          portId: fromPortId
+        },
+        to: {}
+      }
+    }
+  };
 }
 
 export const onLinkMove: IOnLinkMove = ({ linkId, toPosition }) => (chart: IChart): IChart => {
-  const link = chart.links[linkId]
-  link.to.position = toPosition
-  chart.links[linkId] = { ...link }
-  return chart
+  const linkToUpdate = chart.links[linkId];
+  return {
+    ...chart,
+    links: {
+      ...chart.links,
+      [linkId]: { ...linkToUpdate, to: { position: toPosition } }
+    }
+  };
 }
 
 export const onLinkComplete: IOnLinkComplete = (props) => {
   const { linkId, fromNodeId, fromPortId, toNodeId, toPortId, config = {} } = props
   return (chart: IChart): IChart => {
+    const links = {...chart.links};
     if ((config.validateLink ? config.validateLink({ ...props, chart }) : true) && [fromNodeId, fromPortId].join() !== [toNodeId, toPortId].join()) {
```

Comment:
To resolve the merge conflict, this condition needs to be updated so that it starts with `!config.readonly`:
```ts
    if (!config.readonly && (config.validateLink ? config.validateLink({ ...props, chart }) : true) && [fromNodeId, fromPortId].join() !== [toNodeId, toPortId].join()) {
```